classes/middleware/HasRoles.php

```diff
@@ -45,18 +47,31 @@ class HasRoles
     public function handle(Request $request, Closure $next, string $matchableRoles, int $rolesMatchingCriteria = HasRoles::ROLES_MATCH_LOOSE)
     {
         $user = $request->user(); /** @var \PKP\user\User $user */
-        $context = $request->attributes->get('context'); /** @var \PKP\context\Context $context */
 
-        $userRoles = collect($user->getRoles($context?->getId() ?? Application::SITE_CONTEXT_ID))
-            ->map(fn ($role) => $role->getId())
-            ->sort();
+        // Get all user roles.
+        $roleDao = DAORegistry::getDAO('RoleDAO'); /** @var RoleDAO $roleDao */
+        $userRoles = collect( $roleDao->getByUserIdGroupedByContext($user->getId()));
+        $context = $request->attributes->get('context'); /** @var \PKP\context\Context $context */
 
         $matchableRoles = Str::of($matchableRoles)
             ->explode('|')
             ->map(fn ($role) => (int)$role)
             ->sort();
 
-        $matchedRoles = $userRoles->intersect($matchableRoles);
+        $matchedRoles = collect([]);
+
+        // Check for a match amongst Context specific roles
+        if ($context->getId() && $userRoles->has($context->getId())) {
```

Comment:
calling `getId()` on non-contextual request will cause exception like `Error: Call to a member function getId() on null`  as context not available for such request . For example, try visiting the `View Jobs` or `View Failed Jobs` from `Administration` section .